File: kotlin-node/karakum/src/jsMain/resources/plugins/convertSkippedGenerics.js
import ts from "typescript";

export default function (node, context, render) {
    const sourceFileName = node.getSourceFile()?.fileName ?? "generated.d.ts"

    if (
        ts.isTypeReferenceNode(node)
        && ts.isIdentifier(node.typeName)
        && node.typeName.text === "ReadableStream"
        && !node.typeArguments
        && sourceFileName.endsWith("fs/promises.d.ts")
    ) {
        return `${render(node.typeName)}<*>`
    }

    if (
        ts.isTypeReferenceNode(node)
        && ts.isQualifiedName(node.typeName)
        && ts.isIdentifier(node.typeName.left)
        && node.typeName.left.text === "streamWeb"
        && (
            node.typeName.right.text === "ReadableStream"
            || node.typeName.right.text === "WritableStream"
        )
        && !node.typeArguments
        && sourceFileName.endsWith("stream.d.ts")
    ) {
        return `${render(node.typeName)}<*>`
    }

    if (
        ts.isTypeReferenceNode(node)
        && ts.isIdentifier(node.typeName)
        && node.typeName.text === "Channel"
        && !node.typeArguments
        && sourceFileName.endsWith("diagnostics_channel.d.ts")
    ) {
        return `${render(node.typeName)}<*, *>`
    }

    if (
        ts.isExpressionWithTypeArguments(node)
        && ts.isIdentifier(node.expression)
        && node.expression.text === "TracingChannelCollection"
        && !node.typeArguments
        && sourceFileName.endsWith("diagnostics_channel.d.ts")
    ) {
        return `${render(node.expression)}<StoreType, ContextType>`
    }

    if (
        ts.isTypeReferenceNode(node)
        && ts.isQualifiedName(node.typeName)
        && (
            node.typeName.right.text === "TypedArray"
        )
        && !node.typeArguments
    ) {
        return `${render(node.typeName)}<*, *, *, *>`
    }

    if (
        ts.isTypeReferenceNode(node)
        && ts.isQualifiedName(node.typeName)
        && (
            node.typeName.right.text === "ArrayBufferView"
        )
        && !node.typeArguments
    ) {
        return `${render(node.typeName)}<*>`
    }

    if (
        ts.isTypeReferenceNode(node)
        && ts.isIdentifier(node.typeName)
        && (
            node.typeName.text === "DataView"
            || node.typeName.text === "Float32Array"
            || node.typeName.text === "Float64Array"
            || node.typeName.text === "Int8Array"
            || node.typeName.text === "Int16Array"
            || node.typeName.text === "Int32Array"
            || node.typeName.text === "Uint8Array"
            || node.typeName.text === "Uint8ClampedArray"
            || node.typeName.text === "Uint16Array"
            || node.typeName.text === "Uint32Array"
            || node.typeName.text === "BigInt64Array"
            || node.typeName.text === "BigUint64Array"
        )
        && !node.typeArguments
    ) {
        return `${render(node.typeName)}<*>`
    }

    if (
        ts.isExpressionWithTypeArguments(node)
        && ts.isIdentifier(node.expression)
        && node.expression.text === "Uint8Array"
        && !node.typeArguments
    ) {
        return `${render(node.expression)}<ArrayBufferLike>`
    }

    if (
        ts.isTypeReferenceNode(node)
        && ts.isIdentifier(node.typeName)
        && (
            node.typeName.text === "Function"
        )
        && !node.typeArguments
    ) {
        if (ts.isIntersectionTypeNode(node.parent)) {
            return "js.function.JsFunction<Nothing, Any?>"
        }

        if (
            ts.isParameter(node.parent)
            && ts.isIdentifier(node.parent.name)
            && node.parent.name.text === "listener"
        ) {
            return `${render(node.typeName)}<Unit>`
        }

        return `${render(node.typeName)}<*>`
    }

    if (
        ts.isTypeReferenceNode(node)
        && ts.isIdentifier(node.typeName)
        && (
            node.typeName.text === "GeneratorFunction"
            || node.typeName.text === "AsyncGeneratorFunction"
        )
        && !node.typeArguments
    ) {
        return `${render(node.typeName)}<*, *>`
    }

    if (
        ts.isTypeReferenceNode(node)
        && ts.isIdentifier(node.typeName)
        && (
            node.typeName.text === "Generator"
        )
        && !node.typeArguments
    ) {
        return `${render(node.typeName)}<*, *, *>`
    }

    if (
        ts.isTypeReferenceNode(node)
        && ts.isIdentifier(node.typeName)
        && (
            node.typeName.text === "AsyncGenerator"
        )
        && node.typeArguments
        && node.typeArguments.length === 2
    ) {
        const typeArguments = node.typeArguments
            ?.map(typeArgument => render(typeArgument))
            ?.join(", ")

        return `${render(node.typeName)}<${typeArguments}, *>`
    }

    if (
        ts.isTypeReferenceNode(node)
        && ts.isIdentifier(node.typeName)
        && (
            node.typeName.text === "MockFunctionCall"
        )
        && node.typeArguments
        && node.typeArguments.length === 1
    ) {
        const typeArguments = node.typeArguments
            ?.map(typeArgument => render(typeArgument))
            ?.join(", ")

        return `${render(node.typeName)}<${typeArguments}, *, *>`
    }

    if (
        ts.isTypeReferenceNode(node)
        && ts.isIdentifier(node.typeName)
        && node.typeName.text === "Http2ServerResponse"
        && !node.typeArguments
    ) {
        return `${render(node.typeName)}<*>`
    }

    if (
        ts.isTypeReferenceNode(node)
        && ts.isIdentifier(node.typeName)
        && (
            node.typeName.text === "Http2Server"
            || node.typeName.text === "Http2SecureServer"
        )
        && !node.typeArguments
    ) {
        return `${render(node.typeName)}<*, *, *, *>`
    }

    if (
        ts.isExpressionWithTypeArguments(node)
        && ts.isIdentifier(node.expression)
        && node.expression.text === "OutgoingMessage"
        && !node.typeArguments
    ) {
        return `${render(node.expression)}<IncomingMessage>`
    }

    if (
        ts.isExpressionWithTypeArguments(node)
        && ts.isIdentifier(node.expression)
        && node.expression.text === "Function"
        && !node.typeArguments
    ) {
        return "js.function.JsFunction<Nothing, Any?>"
    }

    return null
}

File: kotlin-node/karakum/src/jsMain/kotlin/node/karakum/plugins/convertSkippedGenerics.kt
package node.karakum.plugins

import node.karakum.util.nullable
import io.github.sgrishchenko.karakum.extension.createPlugin
import io.github.sgrishchenko.karakum.util.getParentOrNull
import io.github.sgrishchenko.karakum.util.getSourceFileOrNull
import typescript.asArray
import typescript.isExpressionWithTypeArguments
import typescript.isIdentifier
import typescript.isIntersectionTypeNode
import typescript.isParameter
import typescript.isPropertyAccessExpression
import typescript.isQualifiedName
import typescript.isTypeReferenceNode

val convertSkippedGenerics = createPlugin { node, _, render ->
    nullable {
        val sourceFileName = ensureNotNull(node.getSourceFileOrNull()).fileName

        nullable {
            ensure(sourceFileName.endsWith("fs/promises.d.ts"))

            ensure(isTypeReferenceNode(node))

            val typeName = node.typeName
            ensure(isIdentifier(typeName))
            ensure(typeName.text == "ReadableStream")
            ensure(node.typeArguments == null)

            "${render(typeName)}<*>"
        } ?: nullable {
            ensure(sourceFileName.endsWith("stream.d.ts"))

            ensure(isTypeReferenceNode(node))

            val typeName = node.typeName
            ensure(isQualifiedName(typeName))

            val left = typeName.left
            ensure(isIdentifier(left))
            ensure(left.text == "streamWeb")

            ensure(
                typeName.right.text == "ReadableStream"
                        || typeName.right.text == "WritableStream"
                        || typeName.right.text == "QueuingStrategy"
            )

            ensure(node.typeArguments == null)

            "${render(typeName)}<*>"
        } ?: nullable {
            ensure(sourceFileName.endsWith("stream/consumers.d.ts"))

            ensure(isTypeReferenceNode(node))

            val typeName = node.typeName
            ensure(isIdentifier(typeName))
            ensure(typeName.text == "WebReadableStream")

            ensure(node.typeArguments == null)

            "${render(typeName)}<*>"
        } ?: nullable {
            ensure(sourceFileName.endsWith("diagnostics_channel.d.ts"))

            ensure(isTypeReferenceNode(node))

            val typeName = node.typeName
            ensure(isIdentifier(typeName))
            ensure(typeName.text == "Channel")

            ensure(node.typeArguments == null)

            "${render(typeName)}<*, *>"
        } ?: nullable {
            ensure(sourceFileName.endsWith("diagnostics_channel.d.ts"))

            ensure(isExpressionWithTypeArguments(node))

            val expression = node.expression
            ensure(isIdentifier(expression))
            ensure(expression.text == "TracingChannelCollection")

            ensure(node.typeArguments == null)

            "${render(expression)}<StoreType, ContextType>"
        } ?: nullable {
            ensure(isTypeReferenceNode(node))

            val typeName = node.typeName
            ensure(isQualifiedName(typeName))
            ensure(typeName.right.text == "TypedArray")

            ensure(node.typeArguments == null)

            "${render(typeName)}<*, *, *, *>"
        } ?: nullable {
            ensure(isTypeReferenceNode(node))

            val typeName = node.typeName
            ensure(isQualifiedName(typeName))
            ensure(typeName.right.text == "ArrayBufferView")

            ensure(node.typeArguments == null)

            "${render(typeName)}<*>"
        } ?: nullable {
            ensure(isTypeReferenceNode(node))

            val typeName = node.typeName
            ensure(isIdentifier(typeName))
            ensure(
                typeName.text == "DataView"
                    || typeName.text == "Float32Array"
                    || typeName.text == "Float64Array"
                    || typeName.text == "Int8Array"
                    || typeName.text == "Int16Array"
                    || typeName.text == "Int32Array"
                    || typeName.text == "Uint8Array"
                    || typeName.text == "Uint8ClampedArray"
                    || typeName.text == "Uint16Array"
                    || typeName.text == "Uint32Array"
                    || typeName.text == "BigInt64Array"
                    || typeName.text == "BigUint64Array"
            )

            ensure(node.typeArguments == null)

            "${render(typeName)}<*>"
        } ?: nullable {
            ensure(isExpressionWithTypeArguments(node))

            val expression = node.expression
            ensure(isIdentifier(expression))
            ensure(expression.text == "Uint8Array")

            ensure(node.typeArguments == null)

            "${render(expression)}<ArrayBufferLike>"
        } ?: nullable {
            ensure(isTypeReferenceNode(node))

            val typeName = node.typeName
            ensure(isIdentifier(typeName))
            ensure(typeName.text == "Function")

            ensure(node.typeArguments == null)

            nullable {
                val intersection = ensureNotNull(node.getParentOrNull())
                ensure(isIntersectionTypeNode(intersection))

                "js.function.JsFunction<Nothing, Any?>"
            } ?: nullable {
                val parameter = ensureNotNull(node.getParentOrNull())
                ensure(isParameter(parameter))

                val parameterName = parameter.name
                ensure(isIdentifier(parameterName))
                ensure(parameterName.text == "listener")

                "${render(typeName)}<Unit>"
            } ?: run {
                "${render(typeName)}<*>"
            }
        } ?: nullable {
            ensure(isTypeReferenceNode(node))

            val typeName = node.typeName
            ensure(isIdentifier(typeName))
            ensure(
                typeName.text == "GeneratorFunction"
                        || typeName.text == "AsyncGeneratorFunction"
            )

            ensure(node.typeArguments == null)

            "${render(typeName)}<*, *>"
        } ?: nullable {
            ensure(isTypeReferenceNode(node))

            val typeName = node.typeName
            ensure(isIdentifier(typeName))
            ensure(typeName.text == "Generator")

            ensure(node.typeArguments == null)

            "${render(typeName)}<*, *, *>"
        } ?: nullable {
            ensure(isTypeReferenceNode(node))

            val typeName = node.typeName
            ensure(isIdentifier(typeName))
            ensure(typeName.text == "AsyncGenerator")

            ensure(node.typeArguments?.asArray()?.size == 2)

            val typeArguments = node.typeArguments?.asArray()?.joinToString(", ") { render(it) }

            "${render(typeName)}<${typeArguments}, *>"
        } ?: nullable {
            ensure(isTypeReferenceNode(node))

            val typeName = node.typeName
            ensure(isIdentifier(typeName))
            ensure(typeName.text == "MockFunctionCall")

            ensure(node.typeArguments?.asArray()?.size == 1)

            val typeArguments = node.typeArguments?.asArray()?.joinToString(", ") { render(it) }

            "${render(typeName)}<${typeArguments}, *, *>"
        } ?: nullable {
            ensure(isTypeReferenceNode(node))

            val typeName = node.typeName
            ensure(isIdentifier(typeName))
            ensure(typeName.text == "Http2ServerResponse")

            ensure(node.typeArguments == null)

            "${render(typeName)}<*>"
        } ?: nullable {
            ensure(isTypeReferenceNode(node))

            val typeName = node.typeName
            ensure(isIdentifier(typeName))
            ensure(
                typeName.text == "Http2Server"
                        || typeName.text == "Http2SecureServer"
            )

            ensure(node.typeArguments == null)

            "${render(typeName)}<*, *, *, *>"
        } ?: nullable {
            ensure(isTypeReferenceNode(node))

            val typeName = node.typeName
            ensure(isIdentifier(typeName))
            ensure(
                typeName.text == "ReadableOptions"
                        || typeName.text == "WritableOptions"
                        || typeName.text == "DuplexOptions"
                        || typeName.text == "TransformOptions"
            )

            ensure(node.typeArguments == null)

            "${render(typeName)}<*>"
        } ?: nullable {
            ensure(isTypeReferenceNode(node))

            val typeName = node.typeName
            ensure(isQualifiedName(typeName))

            val left = typeName.left
            ensure(isIdentifier(left))
            ensure(left.text == "stream")

            ensure(
                typeName.right.text == "ReadableOptions"
                        || typeName.right.text == "WritableOptions"
                        || typeName.right.text == "TransformOptions"
            )

            ensure(node.typeArguments == null)

            "${render(typeName)}<*>"
        } ?: nullable {
            ensure(isTypeReferenceNode(node))

            val typeName = node.typeName
            ensure(isIdentifier(typeName))
            ensure(typeName.text == "Dirent")

            ensure(node.typeArguments == null)

            "${render(typeName)}<*>"
        } ?: nullable {
            ensure(sourceFileName.endsWith("crypto.d.ts"))

            ensure(isExpressionWithTypeArguments(node))

            val expression = node.expression
            ensure(isPropertyAccessExpression(expression))

            val expressionExpression = expression.expression
            ensure(isIdentifier(expressionExpression))
            ensure(expressionExpression.text == "stream")

            val expressionName = expression.name
            ensure(isIdentifier(expressionName))
            ensure(expressionName.text == "TransformOptions")

            ensure(node.typeArguments == null)

            "${render(expression)}<node.stream.Transform>"
        } ?: nullable {
            ensure(isExpressionWithTypeArguments(node))

            val expression = node.expression
            ensure(isIdentifier(expression))
            ensure(expression.text == "OutgoingMessage")

            ensure(node.typeArguments == null)

            "${render(expression)}<IncomingMessage>"
        } ?: nullable {
            ensure(isExpressionWithTypeArguments(node))

            val expression = node.expression
            ensure(isIdentifier(expression))
            ensure(expression.text == "Function")

            ensure(node.typeArguments == null)

            "js.function.JsFunction<Nothing, Any?>"
        }
    }
}
